### ISSUE TITLE
Follow-izing many helm contexts

### ISSUE THREAD

**Issue by quicknir** [syl20bnr/spacemacs#6592]
Hi, I recently put in a pull request, #6561, with the idea of using swoop instead of helm-ag for the multi-buffer search. This seemed to be a popular idea. In the end, after testing it with 20+ buffers, I think the performance is a bit of a concern. However, I think we can improve the usability of helm-ag and many other forms of helm. One of (if not the biggest) features of swoop is that it starts off in follow mode automatically. It seems like having to hit `C-c C-f` would not be a big deal in helm-ag, but in practice it is.

It is possible to take any helm function that does not start in follow mode by default, and automatically create a version that does start in follow mode:

```
  (defun followize (helm-source helm-command)
    (lexical-let ((hs helm-source)
                  (hc helm-command))
      (lambda ()
        (interactive)
        (let ((prev-follow-val (helm-get-attribute-from-source-type 'follow hs)))
          (helm-attrset 'follow 1 hs)
          (call-interactively hc)
          (helm-attrset 'follow prev-follow-val hs)
          )
        )
      )
    )
;; usage
  (evil-leader/set-key-for-mode 'c++-mode "q" (followize helm-source-do-ag 'helm-do-ag-buffers))
```

My idea would be to change certain helm invocations to automatically start in follow mode. There could be an option for this in .spacemacs, spacemacs-some-helms-always-follow or something like that, and we could discuss which helm invocations it should affect. The main idea though is to only use it in places where there's zero negative effects. For instance, using this on project ag would be bad, because entries in the list are often files are disk that are not yet in buffers. Activating follow mode there would be slow, and would lead to tons of buffers being opened.

Buffer helm-ag is an obvious one where there's basically no downside, as the files are already open. `SPC b b` is a bit of a mixed bag; for buffers this seems great but recent files are also here, though 95% of my usage is buffers. This seems like a very convenient way to browse buffers to me; it basically combines the advantages of the buffer list and some of the advantages of the buffer micro state. `SPC j i/I` (in develop), that is the imenu jump in buffer/all buffers I think would be a good candidate for this. I think those are the main ones to consider for now, of course the buffer helm-ag search is also applicable to the other searches pt, grep, ack.

Thoughts? If people like this idea I can put together a pull request very quickly.

Note that some people might suggest that the way to solve this is with  helm-follow-mode-persistent. Unfortunately this persists (AFAIK) the value of follow mode across completely different types of helm sessions, so it doesn't really solve the issue.

**Comment by quicknir** [comment]
In addition, with helm-ag in follow mode automatically for certain things, swoop can also be replaced with helm-ag or whatever variants. You lose nothing, and gain performance, and ditch a dependency.

**Comment by deb0ch** [comment]
> For instance, using this on project ag would be bad, because entries in the list are often files are disk that are not yet in buffers. Activating follow mode there would be slow, and would lead to tons of buffers being opened.

Ranger does that and I never found it problematic. As long as these buffers are temporary or as you kill them after I think it would be a nice feature for multi files searches.
Maybe for large files it could be a concern, but not even sure.
Configuration variable for following in large files ?

Another advantage of swoop (imho) is the better readability of its output.

Helm-ag prefixes each occurence with the file pathname, which pollutes the output and make it take a larger portion of the screen than necessary (I'm using Helm on the side of the screen).
Swoop is smarter that way, as it displays the file pathnames as titles above the occurrences and only once per file, which makes the output lighter and more narrow.

**Comment by quicknir** [comment]
@deb0ch Swoop has this advantage only for multi-buffer searches, for single buffer searches helm-ag does the nice thing for single buffer. Unfortunately, the other variants (ack/pt/grep) don't seem to have a single file search? For multi buffer swoop is too slow to switch to in any case though, so while I agree with you about the output, I don't think it's a deciding factor.

Unfortunately I did not look at non-ag options; I didn't realize they were so limited. It may be hard to drop swoop for this reason. But can still make multi buffer searches and imenu follow automatically. I think follow mode for files on disk will be more controversial, and it's also a non-trivial amount of work to get some of the ideas you suggested to work with it, which I won't be tackling in this PR.

**Comment by TheBB** [comment]
> Unfortunately, the other variants (ack/pt/grep) don't seem to have a single file search?

ack and pt are just different search tools, they're handled by the Emacs package helm-ag. The only one that's different is grep.

**Comment by quicknir** [comment]
@TheBB point taken. In that case it may still be possible then. I'll look at putting together a PR this weekend. I'm curious how much attention/use/love actual grep searching gets; I just tried it for the first time and the color scheme with spacemacs light is such that the searched-for text is nearly unreadable. So probably almost everyone is using one of the more advanced tools.

**Comment by CestDiego** [comment]
I just have `C-n` binded to next item and `C-s-n` to next item and follow in help.

**Comment by quicknir** [comment]
@CestDiego This is not the same as swoop's follow mood; when you type in swoop and narrow it does it automatically jumps to the first entry. What you are describing (and what follow mode in helm-ag does) is only active when you press down up. In addition, as I've said, I think following makes sense in certain helm modes but not in others, I don't want to think about which key to hit.

More generally: I put this on hold for a bit because of a recent issue I raised on helm, that has been resolved. It's actually related to following the first entry automatically. I havent' dropped this, just waiting for helm to get updated on melpa before I keep pushing it.
